src/ButtonGroup.php

```diff
@@ -4,143 +4,176 @@
      */
-    public function options(array $value): self
+    public function id(bool|string $value): self
     {
         $new = clone $this;
-        $new->options = $value;
+        $new->id = $value;
 
         return $new;
     }
 
     /**
-     * Generates the buttons that compound the group as specified on {@see buttons}.
-     *
-     * @throws InvalidConfigException
+     * Sets the button group size to be large.
      *
-     * @return string the rendering result.
+     * @return self A new instance of the current class with the button as a large button.
      */
-    private function renderButtons(): string
+    public function large(): self
     {
-        $buttons = [];
+        $new = clone $this;
+        $new->cssClass = 'btn-lg';
 
-        foreach ($this->buttons as $button) {
-            if (is_array($button)) {
-                $visible = ArrayHelper::remove($button, 'visible', true);
+        return $new;
+    }
 
-                if ($visible === false) {
-                    continue;
-                }
+    /**
+     * Sets the button group size to be small.
+     *
+     * @return self A new instance of the current class with the button as a small button.
+     */
+    public function small(): self
+    {
+        $new = clone $this;
+        $new->cssClass = 'btn-sm';
 
-                if (!isset($button['encodeLabel'])) {
-                    $button['encodeLabel'] = $this->encodeLabels;
-                }
+        return $new;
+    }
+
+    /**
+     * Sets the button group to be vertical.
+     *
+     * @return self A new instance of the current class with the button group as vertical.
+     */
+    public function vertical(): self
+    {
+        $new = clone $this;
+        $new->cssClass = 'btn-group-vertical';
 
-                if (!isset($button['options']['type'])) {
-                    ArrayHelper::setValueByPath($button, 'options.type', 'button');
-                }
+        return $new;
+    }
 
-                $buttonWidget = Button::widget()
-                    ->label($button['label'])
-                    ->options($button['options']);
+    public function render(): string
+    {
+        $attributes = $this->attributes;
+        $attributes['role'] = 'group';
+        $id = is_string($this->id) && $this->id !== '' ? $this->id : null;
+        $classes = $attributes['class'] ?? null;
+        unset($attributes['class']);
 
-                if ($button['encodeLabel'] === false) {
-                    $buttonWidget = $buttonWidget->withoutEncodeLabels();
-                }
+        Html::addCssClass($attributes, [self::NAME, $classes, $this->cssClass]);
 
-                $buttons[] = $buttonWidget->render();
-            } else {
-                $buttons[] = $button;
-            }
+        if ($this->id === true) {
+            $id = Html::generateId(self::NAME . '-');
         }
 
-        return implode("\n", $buttons);
+        $buttons = implode(PHP_EOL, $this->buttons);
```

Comment:
I'd use `"\n"` instead. No need for `\r\n\`.

Reply:
done.